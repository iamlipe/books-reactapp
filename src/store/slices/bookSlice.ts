import {createSlice, PayloadAction} from '@reduxjs/toolkit';
import {AxiosError} from 'axios';

export interface Book {
  id: string;
  authors: string[];
  title: string;
  imageUrl: string;
  pageCount: number;
  publisher: string;
  published: number;
}

export interface BookDetailsResponse {
  authors: string[];
  title: string;
  description: string;
  pageCount: number;
  category: string;
  imageUrl: string;
  language: string;
  isbn10: string;
  isbn13: string;
  publisher: string;
  published: number;
  id: string;
}

export interface BooksResponse {
  data: Book[];
  page: number;
  totalItems: number;
  totalPages: number;
}

export interface BooksRequest {
  page?: string;
  amount?: string;
  title?: string;
  category?: string;
}

export interface BookDetailsRequest {
  id: string;
}

interface BookState {
  isLoading: boolean;
  error: AxiosError | null;
  statusCode: number | null;

  allBooks: BooksResponse | null;
  detailsBook: BookDetailsResponse | null;
}

const initialState: BookState = {
  isLoading: false,
  error: null,
  statusCode: null,

  allBooks: null,
  detailsBook: null,
};

const bookSlice = createSlice({
  name: 'book',
  initialState,
  reducers: {
    GET_BOOKS: (state, _: PayloadAction<BooksRequest>) => ({
      ...state,
      isLoading: true,
      error: null,
      statusCode: null,
    }),
    GET_BOOKS_SUCCESS: (
      state,
      {
        payload: {data, status},
      }: PayloadAction<{data: BooksResponse; status: number}>,
    ) => ({
      ...state,
      isLoading: false,
      error: null,
      statusCode: status,

      allBooks: data,
    }),
    GET_BOOKS_FAILURE: (state, {payload: {error}}) => ({
      ...state,
      isLoading: false,
      error,
    }),

    GET_DETAILS_BOOK: (state, _: PayloadAction<BookDetailsRequest>) => ({
      ...state,
      isLoading: true,
      error: null,
      statusCode: null,
    }),

    GET_DETAILS_BOOK_SUCCESS: (
      state,
      {
        payload: {data, status},
      }: PayloadAction<{data: BookDetailsResponse; status: number}>,
    ) => ({
      ...state,
      isLoading: false,
      error: null,
      statusCode: status,

      detailsBook: data,
    }),

    GET_DETAILS_BOOK_FAILURE: (state, {payload: {error}}) => ({
      ...state,
      isLoading: false,
      error,
    }),

    REMOVE_DETAILS_BOOK: state => ({
      ...state,

      detailsBook: null,
    }),
  },
});

const {actions, reducer} = bookSlice;

export const bookState = initialState;

export const {
  GET_BOOKS,
  GET_BOOKS_SUCCESS,
  GET_BOOKS_FAILURE,
  GET_DETAILS_BOOK,
  GET_DETAILS_BOOK_SUCCESS,
  GET_DETAILS_BOOK_FAILURE,
  REMOVE_DETAILS_BOOK,
} = actions;

export default reducer;
