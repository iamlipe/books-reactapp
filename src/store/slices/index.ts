import {combineReducers} from '@reduxjs/toolkit';

import user, {userState} from './userSlice';
import book, {bookState} from './bookSlice';

export const globalState = {user: userState, book: bookState};

export const globalReducer = {user, book};

const rootReducer = combineReducers(globalReducer);

export type RootState = ReturnType<typeof rootReducer>;

export default rootReducer;
