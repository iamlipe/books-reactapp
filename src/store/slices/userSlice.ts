import {createSlice, PayloadAction} from '@reduxjs/toolkit';
import {AxiosError} from 'axios';

export interface LoginRequest {
  email: string;
  password: string;
}

export interface LoginResponse {
  id: string;
  name: string;
  birthdate: string;
  gender: string;
  email: string;
}

export interface User {
  id: string;
  name: string;
  email: string;
  authorization: string | undefined;
  refreshToken: string | undefined;
}

interface UserState {
  isLoading: boolean;
  auth: User | null;
  error: AxiosError | null;
  statusCode: number | null;
}

const initialState: UserState = {
  isLoading: false,
  auth: null,
  error: null,
  statusCode: null,
};

const userSlice = createSlice({
  name: 'user',
  initialState,
  reducers: {
    LOGIN: (state, _: PayloadAction<LoginRequest>) => ({
      ...state,
      isLoading: true,
      error: null,
      statusCode: null,
    }),
    LOGIN_SUCCESS: (
      state,
      {payload: {data, status}}: PayloadAction<{data: User; status: number}>,
    ) => ({
      ...state,
      isLoading: false,
      auth: data,
      statusCode: status,
    }),
    LOGIN_FAILURE: (state, {payload: {error}}) => ({
      ...state,
      isLoading: false,
      error,
    }),

    LOGOUT: (state, _: PayloadAction) => ({
      ...state,
      isLoading: true,
      error: null,
    }),
    LOGOUT_SUCCESS: () => ({
      ...initialState,
    }),
    LOGOUT_FAILURE: (state, {payload: {error}}) => ({
      ...state,
      isLoading: false,
      error,
    }),
  },
});

const {actions, reducer} = userSlice;

export const userState = initialState;

export const {
  LOGIN,
  LOGIN_FAILURE,
  LOGIN_SUCCESS,
  LOGOUT,
  LOGOUT_SUCCESS,
  LOGOUT_FAILURE,
} = actions;
export default reducer;
