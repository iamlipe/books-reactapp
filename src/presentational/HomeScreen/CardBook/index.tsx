import React, {memo} from 'react';
import styled from 'styled-components/native';
import {Dimensions} from 'react-native';
import {RFValue} from 'react-native-responsive-fontsize';
import {Book, GET_DETAILS_BOOK} from '@store/slices/bookSlice';

import {useReduxDispatch} from '@hooks/useReduxDispatch';

interface Props {
  item: Book;
}

const {width} = Dimensions.get('window');

const CardBook = ({item}: Props) => {
  const dispatch = useReduxDispatch();

  function handleGetDetailsBook({id}: {id: string}) {
    dispatch(GET_DETAILS_BOOK({id}));
  }

  return (
    <StyledContainer onPress={() => handleGetDetailsBook({id: item.id})}>
      <StyledImage testID="image-book-card" source={{uri: item.imageUrl}} />
      <StyledRow>
        <StyledColumn>
          <StyledTextTitle>
            {item.title.length > 30
              ? `${item.title.substring(0, 30)}...`
              : item.title}
          </StyledTextTitle>
          {item.authors.map((author, index) => (
            <StyledTextAuthor key={index}>
              {item.authors.length - 1 !== index ? `${author},` : author}
            </StyledTextAuthor>
          ))}
        </StyledColumn>

        <StyledColumn>
          <StyledTextInfo>{`${item.pageCount} p√°ginas`}</StyledTextInfo>
          <StyledTextInfo>{`Editora ${item.publisher}`}</StyledTextInfo>
          <StyledTextInfo>{`Publicado em ${item.published}`}</StyledTextInfo>
        </StyledColumn>
      </StyledRow>
    </StyledContainer>
  );
};

const StyledContainer = styled.TouchableOpacity`
  width: 100%;
  height: 160px;

  flex-direction: row;

  background-color: ${({theme}) => theme.colors.WHITE};

  border-radius: 4px;

  padding: 16px;
  margin-bottom: 16px;
`;

const StyledImage = styled.Image`
  width: ${width * 0.9 * 0.28}px;
`;

const StyledRow = styled.View`
  justify-content: space-between;

  padding: 0 16px;
`;

const StyledColumn = styled.View``;

const StyledTextTitle = styled.Text`
  font-family: ${({theme}) => theme.fonts.HEEBO_MEDIUM};
  font-size: ${RFValue(14)}px;

  color: ${({theme}) => theme.colors.TEXT_DARK};
`;

const StyledTextAuthor = styled.Text`
  font-family: ${({theme}) => theme.fonts.HEEBO_REGULAR};
  font-size: ${RFValue(12)}px;

  color: ${({theme}) => theme.colors.TEXT_PRIMARY};
`;

const StyledTextInfo = styled.Text`
  font-family: ${({theme}) => theme.fonts.HEEBO_REGULAR};
  font-size: ${RFValue(12)}px;

  color: ${({theme}) => theme.colors.TEXT_GRAY};
`;

export default memo(CardBook);
