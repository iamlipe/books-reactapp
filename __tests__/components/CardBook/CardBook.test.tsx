import React from 'react';
import {
  renderWithThemeProvider,
  matchSnapshotWithProvider,
} from '@config/functions/Testing';
import {Book} from '@store/slices/bookSlice';

import CardBook from '@presentational/HomeScreen/CardBook';

describe('CardBook', () => {
  const item: Book = {
    id: '1',
    title: 'Mem칩rias P칩stumas de Br치s Cubas',
    authors: ['Machado de Assis'],
    imageUrl:
      'https://images-na.ssl-images-amazon.com/images/I/41ZS6DFo+vL._SX328_BO1,204,203,200_.jpg',
    pageCount: 159,
    publisher: 'Saraiva',
    published: 2019,
  };

  test('should render correctly', () => {
    renderWithThemeProvider(<CardBook item={item} />);
  });

  test('should match snapshot', () => {
    matchSnapshotWithProvider(<CardBook item={item} />);
  });

  test('should render imageBook', () => {
    const cardBook = renderWithThemeProvider(<CardBook item={item} />);

    const imageBook = cardBook.queryByTestId('image-book-card');

    expect(imageBook).toBeTruthy();
  });

  test('should render title', () => {
    const cardBook = renderWithThemeProvider(<CardBook item={item} />);

    const titleCardBook = cardBook.queryByText(
      item.title.length > 30 ? `${item.title.substring(0, 30)}...` : item.title,
    );

    expect(titleCardBook).toBeTruthy();
  });

  test('should render all authors', () => {
    const cardBook = renderWithThemeProvider(<CardBook item={item} />);

    item.authors.forEach(author => {
      const authorCardBook = cardBook.queryByText(author);

      expect(authorCardBook).toBeTruthy();
    });
  });

  test('should render Paperback', () => {
    const cardBook = renderWithThemeProvider(<CardBook item={item} />);

    const paperbackCardBook = cardBook.queryByText(`${item.pageCount} p치ginas`);

    expect(paperbackCardBook).toBeTruthy();
  });

  test('should render publisher', () => {
    const cardBook = renderWithThemeProvider(<CardBook item={item} />);

    const publisherCardBook = cardBook.queryByText(`Editora ${item.publisher}`);

    expect(publisherCardBook).toBeTruthy();
  });

  test('should render published', () => {
    const cardBook = renderWithThemeProvider(<CardBook item={item} />);

    const publishedCardBook = cardBook.queryByText(
      `Publicado em ${item.published}`,
    );

    expect(publishedCardBook).toBeTruthy();
  });
});
