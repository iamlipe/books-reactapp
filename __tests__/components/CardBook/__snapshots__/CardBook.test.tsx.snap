// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardBook should match snapshot 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "borderRadius": 4,
      "flexDirection": "row",
      "height": 160,
      "marginBottom": 16,
      "opacity": 1,
      "paddingBottom": 16,
      "paddingLeft": 16,
      "paddingRight": 16,
      "paddingTop": 16,
      "width": "100%",
    }
  }
>
  <Image
    source={
      Object {
        "uri": "https://images-na.ssl-images-amazon.com/images/I/41ZS6DFo+vL._SX328_BO1,204,203,200_.jpg",
      }
    }
    style={
      Array [
        Object {
          "width": 189.00000000000003,
        },
      ]
    }
    testID="image-book-card"
  />
  <View
    style={
      Array [
        Object {
          "justifyContent": "space-between",
          "paddingBottom": 0,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 0,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#333333",
              "fontFamily": "Heebo-Medium",
              "fontSize": 27,
            },
          ]
        }
      >
        Mem칩rias P칩stumas de Br치s Cuba...
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#AB2680",
              "fontFamily": "Heebo-Regular",
              "fontSize": 24,
            },
          ]
        }
      >
        Machado de Assis
      </Text>
    </View>
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#999999",
              "fontFamily": "Heebo-Regular",
              "fontSize": 24,
            },
          ]
        }
      >
        159 p치ginas
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#999999",
              "fontFamily": "Heebo-Regular",
              "fontSize": 24,
            },
          ]
        }
      >
        Editora Saraiva
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#999999",
              "fontFamily": "Heebo-Regular",
              "fontSize": 24,
            },
          ]
        }
      >
        Publicado em 2019
      </Text>
    </View>
  </View>
</View>
`;
